editor/utils/with-history/index.js

```diff
@@ -14,46 +14,82 @@ import { includes } from 'lodash';
  * @return {Function} Enhanced reducer.
  */
 export default function withHistory( reducer, options = {} ) {
+	const initialPresent = reducer( undefined, {} );
+
 	const initialState = {
-		past: [],
-		present: reducer( undefined, {} ),
+		// Past already contains record of present since changes are buffered in present.
+		past: [ initialPresent ],
+		present: initialPresent,
 		future: [],
 	};
 
+	let lastAction;
+
 	return ( state = initialState, action ) => {
 		const { past, present, future } = state;
+		const previousAction = lastAction;
+		const {
```

Comment:
This code runs on every dispatch, and these variable assignments could just as well occur in the top-level scope (the reducer creator) to avoid unnecessary garbage collection.